core/base/unaligned_access.hpp

```diff
@@ -0,0 +1,389 @@
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_UNALIGNED_ACCESS_HPP_
+#define GKO_CORE_BASE_UNALIGNED_ACCESS_HPP_
+
+
+#include <cstring>
+#include <functional>
+
+
+#include <ginkgo/core/base/types.hpp>
+
+
+namespace gko {
+
+
+/**
+ * Copies the value in the m-th byte of ptr.
+ *
+ * @copy the value in the m-th byte of ptr.
+ */
+template <typename T>
+void set_value_chunk(void *ptr, size_type start, T value)
+{
+    //    std::memcpy(static_cast<std::int8_t *>(ptr) + start, &value,
+    //    sizeof(T)); std::memcpy(ptr + start, &value, sizeof(T));
+    std::memcpy(static_cast<unsigned char *>(ptr) + start, &value, sizeof(T));
+}
+
+
+/**
+ * Returns the value in the m-th byte of ptr, which is adjusting to T class.
+ *
+ * @return the value in the m-th byte of ptr, which is adjusting to T class.
+ */
+template <typename T>
+// T get_value_chunk(const uint8 *ptr, size_type start)
+T get_value_chunk(const void *ptr, size_type start)
+{
+    T val{};
+    //    std::memcpy(&val, ptr + start, sizeof(T));
+    //    std::memcpy(&val, static_cast<std::int8_t *>(ptr) + start, sizeof(T));
+    std::memcpy(&val, static_cast<const unsigned char *>(ptr) + start,
+                sizeof(T));
+    return val;
+}
+
+
+// inline void cnt_next_position(const uint8 ind, size_type &shf)
+inline void cnt_next_position(const size_type colRS, size_type &shf,
```

Comment:
This and all following functions need to be moved to a separate header file since they don't do anything with memory access (see the name of the header file).

Reply:
I have created core/matrix/bccoo_helper.hpp in which these functions are included.